---
title: "Part 5: Isoform switch / Alternative splicing analyses"
format:
  html:
    embed-resources: true
    toc: true
    toc-location: left
    toc-depth: 3
    title-block-banner: "#00A7FF"
css: report_assets/style.css
editor: visual
bibliography: report_assets/references.bib
---

# Isoform switch / Alternative splicing analyses

Note: This workbooks in an adaptation of the official Bioconductor vignette by Kristoffer Vitting-Seerup. Please use the [vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/IsoformSwitchAnalyzeR/inst/doc/IsoformSwitchAnalyzeR.html) for further details or to expand any point.

The ENCODE consortium recently found that on average, \>6 transcripts/alternative isoforms are generated per gene, and even much more in certain genes. Many of these remain uncharacterized, and they may be more common than expected, with important functional consequences. Recent breakthroughs in sequencing techniques and bioinformatics now allow to accurately reconstruct and quantify full-length gene isoforms, either from short RNA-sequencing data via tools such as StringTie, Kallisto and Salmon, or from long-read RNA-seq that can provide us with full length transcripts. To leverage the full potential of RNA-seq data and these full length transcripts, changes in isoform usage must be analysed. Unfortunately, this is rarely done. IsoformSwitchAnalyzeR enables statistical identification of isoform switches from RNA-seq derived quantification of novel and/or known full-length isoforms. Functional consequences and associated changes in alternative splicing can be also predicted.

`IsoformSwitchAnalyzeR` measures isoform usage via isoform fraction (IF) values which quantifies the fraction of the parent gene expression originating from a specific isoform (calculated as isoform_exp / gene_exp). Consequently, the difference in isoform usage is quantified as the difference in isoform fraction (dIF) calculated as IF2 - IF1, and these dIF are used to measure the effect size (like fold changes are in gene/isoform expression analysis).

In short, the aim of `IsoformSwitchAnalyzeR` is to find, annotate and visualise isoform switches with functional consequences, as well as its associated alternative splicing, from full-length RNA-seq derived isoform/transcript level quantification of RNASeq data. Quantification results coming from both short- and long-read sequencing data can be analyzed.

`IsoformSwitchAnalyzeR` performs five high-level tasks:

1.  Statistical identification of isoform switches.

2.  Integration of a wide range of (predicted) annotations for the isoforms involved in the identified switches (e.g. protein domains).

3.  Identification of which isoforms have a predicited functional consequnce (e.g. loss/gain of protein domain).

4.  Visualization of predicted consequences of the isoform switches for individual genes

5.  Analysis of genome wide patterns in both switch consequences and alternative splicing.

![IsoformSwitchAnalyzeRWorkflow](https://training.galaxyproject.org/training-material/topics/transcriptomics/images/differential_isoform/scheme.jpeg)

There are three high-level less-customizable wrapper functions:

1.  `isoformSwitchAnalysisPart1()`
2.  `isoformSwitchAnalysisPart2()`
3.  `isoformSwitchAnalysisCombined()`

## Exercise #1

1.  Can you use the vignette or the help/manual of the functions (i.e. `?isoformSwitchAnalysisPart1()` after loading the package) to summarize what steps they are performing?

::: {.callout-tip collapse="true" appearance="simple"}
## Solution

-   `isoformSwitchAnalysisPart1()` - Identifying isoform switches, annotating ORF, preparing sequences to use external tools if required (check out the [vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/IsoformSwitchAnalyzeR/inst/doc/IsoformSwitchAnalyzeR.html#:~:text=in%20Figure%201.-,Part,-1%20Extract%20Isoform) for details).
-   `isoformSwitchAnalysisPart2()` - Analyzing alternative splicing, predicting functional consequences, plotting.
-   `isoformSwitchAnalysisCombined()` - If one does not plan to incorporate external sequence analysis (or is only interested in splicing), it is possible to run the full workflow using this function.
:::

Alternatively, we are going to be using a step-by-step approach so we can analyze the transcript counts from our previous workbooks and discuss results.

## Importing the data

First, we are loading the package and importing the data:

```{r import}
.rs.restartR(clean=TRUE)
# Installation from the Bioconductor repository:
# First install Bioconductor
# if (!requireNamespace("BiocManager", quietly = TRUE)){
#     install.packages("BiocManager")
#     BiocManager::install()
# }
# Then install the package:
# BiocManager::install("IsoformSwitchAnalyzeR")

# Looading the package:
suppressPackageStartupMessages(library(IsoformSwitchAnalyzeR))

# Importing the data:
# From the vignette:
# salmonQuant <- importIsoformExpression(
#    parentDir = system.file("extdata/",package="IsoformSwitchAnalyzeR")
# )

# There are alternatives, merged samples, different subdirectories, etc! Again, always a good idea to do ?importIsoformExpression and checking out the full vignette.

# Tip: Many packages and vignettes use small subsets of data as examples, which may be included in internal folders such as "extdata". We are using our data here, but when you are on your own, it may be a good idea to double check a package/workflow is working as expected with the test dataset before using real data.
countsDirectoryToUse <- "~/work/gse154927-full/kallisto"

kallistoFilePaths <- list.files(path=countsDirectoryToUse, pattern="*.txt", full.names=T)
kallistoQuant <- importIsoformExpression(sampleVector=kallistoFilePaths)

```

Next, let's take a closer look.Â What have we imported?

```{r import2}
class(kallistoQuant)

head(kallistoQuant$abundance, 5)

head(kallistoQuant$counts, 5)
```

## Exercise #2

1.  What R object is created by the `importIsoformExpression` function?
2.  Even in comprehensive vignettes and workbook there may be obscure concepts. Can you investigate what are the differences between the `abundance` and `counts` imported in `IsoformSwitchAnalyzeR`?

::: {.callout-tip collapse="true" appearance="simple"}
## Solution

1.  A list of with 4 elements. These include 3 data.frames and another list.
2.  A count is the raw count of reads that have overlapped some genomic feature. Abundance would be a more biologically meaningful quantification of the expression (i.e. the biological process) of a gene or transcript, which has been normalized in some way. Counts are better for the statistical analysis but the abundance estimates are better to measuring effect sizes.
:::

What else is needed?

1.  Apart from the Kallisto counts, we will also need some information about the analyses that we want to perform. Similar to the previous session, we need to specify the experimental design matrix, i.e. biologically meaningful comparisons and possible covariates to include in the statistical models if there are unwanted sources of variation not due to experimental conditions you are interested in but that may influence results).
2.  Annotation file with transcript structure, which transcripts are known for which genes, coordinates, etc (i.e. GTF file).
3.  Nucleotide sequence of the transcripts (FASTA file).

```{r import3}

# Matrix design:
myDesign <- data.frame(
  sampleID = colnames(kallistoQuant$abundance)[-1],
  condition = gsub("_[1-4]","",colnames(kallistoQuant$abundance)[-1])
)

# Note that as in differential expression analyses, at least 3 replicates is a must for statistical analyses!

# We also download/locate/prepare:
annotation_file <- "~/work/grch38-p13-gencode-release-42/annotation/genes_protein_coding.gtf"
transcript_sequences <- "~/work/grch38-p13-gencode-release-42/annotation/genes_protein_coding_kallistoIndex/transcripts.fa"
myComparisons <- data.frame(
  condition_1 = c("HCT116_EpCAMhigh", "HCT116_EpCAMlow"),
  condition_2 = c("SW480_EpCAMhigh", "SW480_EpCAMlow")
)
# Note that the quantification that we are importing, annotation file, transcript sequences... must of course share the transcript IDs! Be careful when importing databases! Alternatively, the full vignette contain details on how to analyse de novo isoforms/transcripts.

# An object called switchAnalyzeRlist is going to integrate all of this
aSwitchList <- importRdata(
  isoformCountMatrix = kallistoQuant$counts,
  isoformRepExpression = kallistoQuant$abundance,
  designMatrix = myDesign,
  comparisonsToMake = myComparisons,
  isoformExonAnnoation = annotation_file,
  isoformNtFasta = transcript_sequences   
)

summary(aSwitchList)
```

## Exercise #3

1.  Check out the resources used! How much GB of RAM is your R session using? In real analyses, the availability of computational resources is often a problem and something to keep in mind!
2.  How many isoforms have been quantified and how many genes are represented?
3.  How many isoforms were not expressed in any sample?

::: {.callout-tip collapse="true" appearance="simple"}
## Solution

1.  It will depend on whether this is a new session, whether you have kept old objects, have freed unused objects, etc... but you may notice that the objects in this analyses are bigger than in the previous ones!
2.  148,137 isoforms from 18,732 genes
3.  21,265 isoforms
:::

## Filtering

Your `switchAnalyzeRlist` will most likely contain genes and isoforms that are irrelevant and should be filtered out. The function in charge of this is `preFilter`.

```{r filter}
aSwitchList <- preFilter(
  switchAnalyzeRlist = aSwitchList,
  geneExpressionCutoff = 1,
  isoformExpressionCutoff = 0,
  removeSingleIsoformGenes = TRUE
)

```

## Exercise #4

1.  Check out the manual of the function documentation (help) and the full vignette if required. Which aspects could your filter on?
2.  Can you think about any other way of filtering?
3.  How many transcripts were removed?

::: {.callout-tip collapse="true" appearance="simple"}
## Solution

1.  Multi-isoform genes, gene expression, isoform expression, isoform fraction (isoform usage), unwanted isoform classes, unwanted gene biotypes, genes without differential isoform usage.
2.  There may be non-used isoforms that only appear here because they were in the isoform/gene annotation used... We could filter here, or remove them from the annotation in previous steps, so they are not even quantified!
3.  More than 40% of them! 61,037 transcripts.
:::

## Computing Isoform Switches

There are different approaches supported in `IsoformSwitchAnalyzeR`. We are using `DEXSeq`. This is a step that would take long and exhaust the computational resources of this session. We are then subsampling and taking only a subset of our data:

```{r dexseq}
# Subsampling, so we only get results with higher differential isoform fraction and the object is not that big anymore
aSwitchList_sub <- subsetSwitchAnalyzeRlist(
  switchAnalyzeRlist = aSwitchList,
  subset = abs(aSwitchList$isoformFeatures$dIF) > 0.15
)

# We can also clean up the old one!
rm(aSwitchList); rm(kallistoQuant); gc()

# Note that we need to think about two parameters before computing switches:
alpha <- 0.05 # FDR corrected P-value (Q-value) cutoff
dIFcutoff <- 0.10 #  minimum (absolute) change in isoform usage (dIF) to be considered (effect size)

aSwitchList_sub <- isoformSwitchTestDEXSeq(
  switchAnalyzeRlist = aSwitchList_sub,
  reduceToSwitchingGenes=FALSE,
  alpha=alpha,
  dIFcutoff = dIFcutoff,
  onlySigIsoforms = FALSE
)

# Get some numbers
extractSwitchSummary(aSwitchList_sub)

table_top_10 <- extractTopSwitches(
  aSwitchList_sub,
  filterForConsequences = FALSE,
  n=10,
  inEachComparison = TRUE
)

table_top_10

extractSwitchOverlap(
  aSwitchList_sub,
  filterForConsequences=FALSE,
  plotIsoforms = FALSE
)

# Get nice Volcano plots to visualise differences
datToPlot <- aSwitchList_sub$isoformFeatures
datToPlot$isoform_switch_q_value <- ifelse(datToPlot$isoform_switch_q_value < minPValToPlot, minPValToPlot, datToPlot$isoform_switch_q_value)
minPValToPlot <- 1E-50

ggplot(data=datToPlot, aes(x=dIF, y=-log10(isoform_switch_q_value))) +
  geom_point(
    aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
    size=1
  ) +
  geom_hline(yintercept = -log10(0.05), linetype='dashed') + # default cutoff
  geom_vline(xintercept = c(-0.1, 0.1), linetype='dashed') + # default cutoff
  facet_wrap( ~ condition_1 ~ condition_2) +
  #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
  scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
  labs(x='dIF', y='-Log10 ( Isoform Switch Q Value )', 
       caption=sprintf("(Q Values set to a maximum of %s)", -log10(minPValToPlot))) +
  theme_bw()

# Take into account the gene-level fold change too
ggplot(data=aSwitchList_sub$isoformFeatures, aes(x=gene_log2_fold_change, y=dIF)) +
  geom_point(
    aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
    size=1
  ) + 
  facet_wrap( ~ condition_1 ~ condition_2) +
  #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
  geom_hline(yintercept = 0, linetype='dashed') +
  geom_vline(xintercept = 0, linetype='dashed') +
  scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
  labs(x='Gene log2 fold change', y='dIF') +
  theme_bw()


```

## Exercise #5

1.  How many isoform switches have we detected in total? Corresponding to how many genes?
2.  How many gene comparisons were tested?
3.  Which is the gene with the switch considered to be top?
4.  How many overlapping genes are shared between comparisons?

::: {.callout-tip collapse="true" appearance="simple"}
## Solution

1.  1,414 switches involving 2,157 isoforms of 1,039 genes
2.  Isoform switch analysis was performed for 5,117 gene comparisons
3.  GUSB
4.  216 genes with 216 switches
:::

## Computing alternative splicing

Since the gene/exon structure is known, alternative splicing can be identified. In this package, alternative splicing is used as a term which covers: Alternative Splicing (AS), Alternative Transcription Start Sites (ATSS, sometimes called alternative first exon), and Alternative Transcription Termination Sites (ATTS, sometimes called alternative last exon).

```{r splic}

aSwitchList_sub <- analyzeAlternativeSplicing(
  switchAnalyzeRlist = aSwitchList_sub
)

head(aSwitchList_sub$AlternativeSplicingAnalysis, 10)

extractSplicingSummary(aSwitchList_sub)

extractSplicingEnrichment(aSwitchList_sub,splicingToAnalyze='all')
```

## Exercise #6

1.  Which splicing events can we capture? (Hint: see column names of `aSwitchList_sub$AlternativeSplicingAnalysis`, `?analyzeAlternativeSplicing`, labels of the plot...
2.  Table manipulations (of data.frames matrices) are crucial when working with omics datasets and interrogating results in R. Choose two splicing events and report how many have we detected checking out the table results.
3.  What would the `onlySwitchingGenes` parameter in the `analyzeAlternativeSplicing` function do? Is there anything unusual in the manual page?
4.  Is there type of splicing event significantly enriched?

## Solution

::: {.callout-tip collapse="true" appearance="simple"}
1.  ES, MEE, MES, IR, A5, A3, ATSS, ATTS
2.  Will depend on your choice, but one of:
-   ES-793

-   MEE-20

-   MES-208

-   IR-167

-   A5-387

-   A3-372

-   ATSS-1,485

-   ATTS- 1,433
3.  Only genes with isoform switches would be included (filtering by the parameters `alpha` and `dIFcutoff`). Contradictory default values are mentioned? But it's `TRUE`!
4.  Yes! IR in the second comparison
:::

## Predicting potential consequences

Changes in isoform usage may have important functional consequences. Depending on the information that we have gathered in the workflow above, can we try and predict some consequences?

```{r consequen}
consequencesOfInterest <- c("intron_retention") # The only thing that we can compute

aSwitchList_sub <- analyzeSwitchConsequences(aSwitchList_sub, consequencesToAnalyze = consequencesOfInterest, dIFcutoff = 0.4)

extractSwitchSummary(aSwitchList_sub, dIFcutoff = 0.4, filterForConsequences = TRUE)

extractSwitchSummary(aSwitchList_sub, dIFcutoff = 0.4, filterForConsequences = FALSE)
```

## Exercise #7

1.  Why can we only choose intron_retention here?
2.  How many genes with isoform switches have potential functional consequences based on the differences in intron retention?

## Solution

::: {.callout-tip collapse="true" appearance="simple"}
1.  Because for the sake of time we skipped some steps were we would have added more information to the `switchAnalyzeRlist`, executing external software. These could have included others aspects such as coding potential, protein domains, prediction of intrinsically disordered regions, etc.
2.  7 out of 221 genes with switches with a dIF \> 0.4.
:::

## Interrogating particular genes

Articular genes

```{r genes}
switchPlot(
  aSwitchList_sub,
  gene="GUSB",
  condition1 = "SW480_EpCAMhigh",
  condition2 = "HCT116_EpCAMhigh",
  plotTopology=FALSE
)

```
