<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.613">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Part 1: QC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="1_qc_files/libs/clipboard/clipboard.min.js"></script>
<script src="1_qc_files/libs/quarto-html/quarto.js"></script>
<script src="1_qc_files/libs/quarto-html/popper.min.js"></script>
<script src="1_qc_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="1_qc_files/libs/quarto-html/anchor.min.js"></script>
<link href="1_qc_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="1_qc_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="1_qc_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="1_qc_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="1_qc_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<style>
    .quarto-title-block .quarto-title-banner {
      background: #00A7FF;
    }
    </style>


<link rel="stylesheet" href="style.css">
</head>

<body>

<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Part 1: QC</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul class="collapse">
  <li><a href="#downloading-the-data" id="toc-downloading-the-data" class="nav-link" data-scroll-target="#downloading-the-data">Downloading the data</a></li>
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1">Exercise #1</a></li>
  </ul></li>
  <li><a href="#fastqc-multiqc" id="toc-fastqc-multiqc" class="nav-link" data-scroll-target="#fastqc-multiqc">FastQC &amp; MultiQC</a>
  <ul class="collapse">
  <li><a href="#exersise-2" id="toc-exersise-2" class="nav-link" data-scroll-target="#exersise-2">Exersise #2</a></li>
  </ul></li>
  <li><a href="#fastqscreen" id="toc-fastqscreen" class="nav-link" data-scroll-target="#fastqscreen">FastqScreen</a>
  <ul class="collapse">
  <li><a href="#exercise-3" id="toc-exercise-3" class="nav-link" data-scroll-target="#exercise-3">Exercise #3</a></li>
  </ul></li>
  <li><a href="#preprocessing-with-fastp" id="toc-preprocessing-with-fastp" class="nav-link" data-scroll-target="#preprocessing-with-fastp">Preprocessing with <code>fastp</code></a>
  <ul class="collapse">
  <li><a href="#exercise-4" id="toc-exercise-4" class="nav-link" data-scroll-target="#exercise-4">Exercise #4</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>Welcome to our introductory RNA-seq course.</p>
<p>In this part of the course, we will be primarily working with yeast samples, initially sequenced and analyzed at the Functional Genomics Center Z√ºrich. We will be exploring this dataset thoroughly, starting with the QC discussed in the presentations. For this first part, we will be dealing with the full data. For later parts, such as the mapping module, subsets of the data will be used instead to save computational resources and allow for faster processing.</p>
<section id="downloading-the-data" class="level2">
<h2 class="anchored" data-anchor-id="downloading-the-data">Downloading the data</h2>
<p>First, let‚Äôs download the data. The script will download a tar file containing our fastq files, decompress it, and deposit the files into the folder <code>data/tmp/yeast_fastq_full</code>. To execute this script, press on the green arrow button (for all such bash scripts, we first switch to the root directory i.e.&nbsp;<code>/home/rstudio/work/rna-seq-course-feb2023</code> , so the locations of all output files is clearer).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># First we make a directory to house our data</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="va">YEAST_DIR</span><span class="op">=</span>data/tmp/yeast_fastq_full</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$YEAST_DIR</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the data</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$YEAST_DIR</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://fgcz-gstore.uzh.ch/public/RNASeqCourse/yeast_full.tar</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract data</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> yeast_full.tar</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-f</span> yeast_full.tar</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Back home</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="scrolling"><code>  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0 6180M    0 47.2M    0     0   110M      0  0:00:56 --:--:--  0:00:56  109M
  3 6180M    3  194M    0     0   136M      0  0:00:45  0:00:01  0:00:44  136M
  5 6180M    5  329M    0     0   135M      0  0:00:45  0:00:02  0:00:43  135M
  7 6180M    7  482M    0     0   140M      0  0:00:43  0:00:03  0:00:40  140M
 10 6180M   10  644M    0     0   145M      0  0:00:42  0:00:04  0:00:38  145M
 12 6180M   12  772M    0     0   142M      0  0:00:43  0:00:05  0:00:38  145M
 14 6180M   14  915M    0     0   142M      0  0:00:43  0:00:06  0:00:37  144M
 17 6180M   17 1081M    0     0   145M      0  0:00:42  0:00:07  0:00:35  150M
 20 6180M   20 1237M    0     0   146M      0  0:00:42  0:00:08  0:00:34  151M
 22 6180M   22 1388M    0     0   147M      0  0:00:41  0:00:09  0:00:32  148M
 25 6180M   25 1558M    0     0   149M      0  0:00:41  0:00:10  0:00:31  157M
 27 6180M   27 1698M    0     0   148M      0  0:00:41  0:00:11  0:00:30  156M
 29 6180M   29 1823M    0     0   146M      0  0:00:42  0:00:12  0:00:30  148M
 32 6180M   32 1990M    0     0   148M      0  0:00:41  0:00:13  0:00:28  150M
 35 6180M   35 2174M    0     0   150M      0  0:00:41  0:00:14  0:00:27  157M
 37 6180M   37 2336M    0     0   151M      0  0:00:40  0:00:15  0:00:25  155M
 40 6180M   40 2486M    0     0   151M      0  0:00:40  0:00:16  0:00:24  157M
 42 6180M   42 2640M    0     0   151M      0  0:00:40  0:00:17  0:00:23  163M
 45 6180M   45 2808M    0     0   152M      0  0:00:40  0:00:18  0:00:22  163M
 47 6180M   47 2950M    0     0   151M      0  0:00:40  0:00:19  0:00:21  155M
 50 6180M   50 3114M    0     0   152M      0  0:00:40  0:00:20  0:00:20  155M
 52 6180M   52 3270M    0     0   152M      0  0:00:40  0:00:21  0:00:19  156M
 55 6180M   55 3405M    0     0   151M      0  0:00:40  0:00:22  0:00:18  152M
 57 6180M   57 3569M    0     0   152M      0  0:00:40  0:00:23  0:00:17  152M
 60 6180M   60 3736M    0     0   152M      0  0:00:40  0:00:24  0:00:16  157M
 63 6180M   63 3897M    0     0   153M      0  0:00:40  0:00:25  0:00:15  156M
 65 6180M   65 4069M    0     0   153M      0  0:00:40  0:00:26  0:00:14  159M
 68 6180M   68 4220M    0     0   153M      0  0:00:40  0:00:27  0:00:13  162M
 70 6180M   70 4374M    0     0   153M      0  0:00:40  0:00:28  0:00:12  161M
 73 6180M   73 4527M    0     0   153M      0  0:00:40  0:00:29  0:00:11  158M
 75 6180M   75 4695M    0     0   154M      0  0:00:40  0:00:30  0:00:10  159M
 79 6180M   79 4888M    0     0   155M      0  0:00:39  0:00:31  0:00:08  163M
 82 6180M   82 5071M    0     0   156M      0  0:00:39  0:00:32  0:00:07  170M
 84 6180M   84 5247M    0     0   156M      0  0:00:39  0:00:33  0:00:06  174M
 87 6180M   87 5403M    0     0   156M      0  0:00:39  0:00:34  0:00:05  175M
 89 6180M   89 5551M    0     0   156M      0  0:00:39  0:00:35  0:00:04  171M
 92 6180M   92 5695M    0     0   156M      0  0:00:39  0:00:36  0:00:03  161M
 94 6180M   94 5867M    0     0   156M      0  0:00:39  0:00:37  0:00:02  159M
 97 6180M   97 6016M    0     0   156M      0  0:00:39  0:00:38  0:00:01  153M
 99 6180M   99 6157M    0     0   156M      0  0:00:39  0:00:39 --:--:--  150M
100 6180M  100 6180M    0     0   156M      0  0:00:39  0:00:39 --:--:--  150M
20191112.A-G1_R1.fastq.gz
20191112.A-G2_R1.fastq.gz
20191112.A-G3_R1.fastq.gz
20191112.A-G4_R1.fastq.gz
20191112.A-GE1_R1.fastq.gz
20191112.A-GE2_R1.fastq.gz
20191112.A-GE3_R1.fastq.gz
20191112.A-GE4_R1.fastq.gz
dataset.tsv</code></pre>
</div>
</div>
<p>The data should be downloaded into <code>data/tmp/yeast_fastq_full/</code>. To check this, run</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> ../../data/tmp/yeast_fastq_full</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>total 6.1G
-rw-r--r-- 1 rstudio users 938M Jan 18 07:32 20191112.A-G1_R1.fastq.gz
-rw-r--r-- 1 rstudio users 560M Jan 18 07:33 20191112.A-G2_R1.fastq.gz
-rw-r--r-- 1 rstudio users 793M Jan 18 07:36 20191112.A-G3_R1.fastq.gz
-rw-r--r-- 1 rstudio users 748M Jan 18 07:38 20191112.A-G4_R1.fastq.gz
-rw-r--r-- 1 rstudio users 626M Jan 18 07:39 20191112.A-GE1_R1.fastq.gz
-rw-r--r-- 1 rstudio users 491M Jan 18 07:40 20191112.A-GE2_R1.fastq.gz
-rw-r--r-- 1 rstudio users 529M Jan 18 07:42 20191112.A-GE3_R1.fastq.gz
-rw-r--r-- 1 rstudio users 1.5G Jan 18 07:45 20191112.A-GE4_R1.fastq.gz
-rwxr-x--- 1 rstudio users 2.4K Jan 18 13:17 dataset.tsv</code></pre>
</div>
</div>
<p>You should see 9 files in this directory. 8 files ending in fastq.gz and one ‚Äúdataset.tsv‚Äù file. The ‚Äúdataset.tsv‚Äù file gives some details about the samples. Let‚Äôs load this file now in R to see how it looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">"../../data/tmp/yeast_fastq_full/dataset.tsv"</span>) <span class="sc">%&gt;%</span> <span class="fu">as_tibble</span>()</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>ds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 √ó 20
  Name  Condit‚Ä¶¬π Read1‚Ä¶¬≤ Species Fragm‚Ä¶¬≥ Sampl‚Ä¶‚Å¥ Tube ‚Ä¶‚Åµ Sampl‚Ä¶‚Å∂ Plate‚Ä¶‚Å∑ LibCo‚Ä¶‚Å∏
  &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;     &lt;int&gt;   &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;lgl&gt;     &lt;int&gt;
1 G1    Glucose  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
2 G2    Glucose  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
3 G3    Glucose  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
4 G4    Glucose  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
5 GE1   GlycEth  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
6 GE2   GlycEth  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
7 GE3   GlycEth  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
8 GE4   GlycEth  p3005/‚Ä¶ n/a           0     200 p3005_‚Ä¶ bfs_20‚Ä¶ NA            0
# ‚Ä¶ with 10 more variables: `LibConc_qPCR [Characteristic]` &lt;int&gt;,
#   `RIN [Characteristic]` &lt;int&gt;, Adapter1 &lt;chr&gt;, strandMode &lt;chr&gt;,
#   LibraryPrepKit &lt;chr&gt;, EnrichmentMethod &lt;chr&gt;,
#   `InputAmount [Characteristic]` &lt;int&gt;, `Order Id [B-Fabric]` &lt;int&gt;,
#   `PlateName [Characteristic]` &lt;chr&gt;, `Read Count` &lt;int&gt;, and abbreviated
#   variable names ¬π‚Äã`Condition [Factor]`, ¬≤‚Äã`Read1 [File]`,
#   ¬≥‚Äã`FragmentSize [Characteristic]`, ‚Å¥‚Äã`SampleConc [Characteristic]`, ‚Ä¶</code></pre>
</div>
</div>
</section>
<section id="exercise-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1">Exercise #1</h2>
<ol type="1">
<li><p>How many groups of samples are there?</p></li>
<li><p>What Library Prep Kit was used?</p></li>
<li><p>What is the read count? Is this what we expect for RNA-seq?</p></li>
<li><p>Is there information about the strandedness of the libraries?</p></li>
</ol>
<div class="callout-note callout callout-style-simple callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>2 groups: Glucose and GlycEth (i.e.&nbsp;Glucose+Ethanol)</p></li>
<li><p>Lexogen mRNA Stranded Kit</p></li>
<li><p>The samples vary between 5M to just under 20M reads. In general, we seek to get around 20M samples for bulk RNA-seq of human and mouse samples. Given that these are yeast samples and the corresponding genome is significantly smaller, even 5M reads is enough to get good results (see Count QC in chapter 3).</p></li>
<li><p>The libraries are anti-sense.</p></li>
</ol>
</div>
</div>
</div>
</section>
</section>
<section id="fastqc-multiqc" class="level1">
<h1>FastQC &amp; MultiQC</h1>
<p>First, let‚Äôs run FastQC. This tool was developed by Babraham Bioinformatics <span class="citation" data-cites="Andrews:2010tn">(<a href="#ref-Andrews:2010tn" role="doc-biblioref">Andrews 2010</a>)</span> and is what we will primarily use for QC of our fastq files. In addition, we will run MultiQC <span class="citation" data-cites="ewels2016">(<a href="#ref-ewels2016" role="doc-biblioref">Ewels et al. 2016</a>)</span> which aims to aggregate outputs from tools like FastQC into a single report for easier viewing. Let‚Äôs run these tools now. This process takes about 10 minutes on 8 cores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Make directories</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="va">PROCESSED_FASTQC</span><span class="op">=</span>results/1_qc/fastqc</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="va">PROCESSED_MULTIQC</span><span class="op">=</span>results/1_qc/multiqc</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$PROCESSED_FASTQC</span> <span class="va">$PROCESSED_MULTIQC</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Process with FastQC</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="ex">/opt/FastQC/fastqc</span> <span class="at">--extract</span> <span class="at">-o</span> <span class="va">$PROCESSED_FASTQC</span> <span class="dt">\</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">-t</span> 8 <span class="at">-a</span> data/supplementary-files/adapter_list.txt <span class="at">--kmers</span> 7 <span class="dt">\</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  data/tmp/yeast_fastq_full/<span class="pp">*</span>fastq.gz</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine result with multiqc</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="ex">multiqc</span> <span class="at">--outdir</span> <span class="va">$PROCESSED_MULTIQC</span> <span class="va">$PROCESSED_FASTQC</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Back home</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="scrolling"><code>Started analysis of 20191112.A-G1_R1.fastq.gz
Started analysis of 20191112.A-G2_R1.fastq.gz
Started analysis of 20191112.A-G3_R1.fastq.gz
Started analysis of 20191112.A-G4_R1.fastq.gz
Started analysis of 20191112.A-GE1_R1.fastq.gz
Started analysis of 20191112.A-GE2_R1.fastq.gz
Started analysis of 20191112.A-GE3_R1.fastq.gz
Started analysis of 20191112.A-GE4_R1.fastq.gz
Approx 5% complete for 20191112.A-GE2_R1.fastq.gz
Approx 5% complete for 20191112.A-G2_R1.fastq.gz
Approx 5% complete for 20191112.A-GE3_R1.fastq.gz
Approx 5% complete for 20191112.A-GE1_R1.fastq.gz
Approx 5% complete for 20191112.A-G4_R1.fastq.gz
Approx 5% complete for 20191112.A-G3_R1.fastq.gz
Approx 5% complete for 20191112.A-G1_R1.fastq.gz
Approx 10% complete for 20191112.A-GE2_R1.fastq.gz
Approx 10% complete for 20191112.A-G2_R1.fastq.gz
Approx 10% complete for 20191112.A-GE3_R1.fastq.gz
Approx 10% complete for 20191112.A-GE1_R1.fastq.gz
Approx 15% complete for 20191112.A-GE2_R1.fastq.gz
Approx 10% complete for 20191112.A-G4_R1.fastq.gz
Approx 5% complete for 20191112.A-GE4_R1.fastq.gz
Approx 10% complete for 20191112.A-G3_R1.fastq.gz
Approx 15% complete for 20191112.A-GE3_R1.fastq.gz
Approx 15% complete for 20191112.A-G2_R1.fastq.gz
Approx 15% complete for 20191112.A-GE1_R1.fastq.gz
Approx 10% complete for 20191112.A-G1_R1.fastq.gz
Approx 20% complete for 20191112.A-GE2_R1.fastq.gz
Approx 20% complete for 20191112.A-GE3_R1.fastq.gz
Approx 20% complete for 20191112.A-G2_R1.fastq.gz
Approx 15% complete for 20191112.A-G4_R1.fastq.gz
Approx 15% complete for 20191112.A-G3_R1.fastq.gz
Approx 25% complete for 20191112.A-GE2_R1.fastq.gz
Approx 20% complete for 20191112.A-GE1_R1.fastq.gz
Approx 25% complete for 20191112.A-GE3_R1.fastq.gz
Approx 15% complete for 20191112.A-G1_R1.fastq.gz
Approx 25% complete for 20191112.A-G2_R1.fastq.gz
Approx 30% complete for 20191112.A-GE2_R1.fastq.gz
Approx 10% complete for 20191112.A-GE4_R1.fastq.gz
Approx 20% complete for 20191112.A-G4_R1.fastq.gz
Approx 25% complete for 20191112.A-GE1_R1.fastq.gz
Approx 20% complete for 20191112.A-G3_R1.fastq.gz
Approx 30% complete for 20191112.A-GE3_R1.fastq.gz
Approx 30% complete for 20191112.A-G2_R1.fastq.gz
Approx 35% complete for 20191112.A-GE2_R1.fastq.gz
Approx 30% complete for 20191112.A-GE1_R1.fastq.gz
Approx 35% complete for 20191112.A-GE3_R1.fastq.gz
Approx 20% complete for 20191112.A-G1_R1.fastq.gz
Approx 25% complete for 20191112.A-G4_R1.fastq.gz
Approx 40% complete for 20191112.A-GE2_R1.fastq.gz
Approx 25% complete for 20191112.A-G3_R1.fastq.gz
Approx 35% complete for 20191112.A-G2_R1.fastq.gz
Approx 40% complete for 20191112.A-GE3_R1.fastq.gz
Approx 35% complete for 20191112.A-GE1_R1.fastq.gz
Approx 45% complete for 20191112.A-GE2_R1.fastq.gz
Approx 15% complete for 20191112.A-GE4_R1.fastq.gz
Approx 40% complete for 20191112.A-G2_R1.fastq.gz
Approx 30% complete for 20191112.A-G4_R1.fastq.gz
Approx 25% complete for 20191112.A-G1_R1.fastq.gz
Approx 45% complete for 20191112.A-GE3_R1.fastq.gz
Approx 30% complete for 20191112.A-G3_R1.fastq.gz
Approx 50% complete for 20191112.A-GE2_R1.fastq.gz
Approx 40% complete for 20191112.A-GE1_R1.fastq.gz
Approx 45% complete for 20191112.A-G2_R1.fastq.gz
Approx 50% complete for 20191112.A-GE3_R1.fastq.gz
Approx 35% complete for 20191112.A-G4_R1.fastq.gz
Approx 55% complete for 20191112.A-GE2_R1.fastq.gz
Approx 35% complete for 20191112.A-G3_R1.fastq.gz
Approx 30% complete for 20191112.A-G1_R1.fastq.gz
Approx 45% complete for 20191112.A-GE1_R1.fastq.gz
Approx 55% complete for 20191112.A-GE3_R1.fastq.gz
Approx 50% complete for 20191112.A-G2_R1.fastq.gz
Approx 60% complete for 20191112.A-GE2_R1.fastq.gz
Approx 20% complete for 20191112.A-GE4_R1.fastq.gz
Approx 40% complete for 20191112.A-G4_R1.fastq.gz
Approx 50% complete for 20191112.A-GE1_R1.fastq.gz
Approx 60% complete for 20191112.A-GE3_R1.fastq.gz
Approx 40% complete for 20191112.A-G3_R1.fastq.gz
Approx 65% complete for 20191112.A-GE2_R1.fastq.gz
Approx 55% complete for 20191112.A-G2_R1.fastq.gz
Approx 35% complete for 20191112.A-G1_R1.fastq.gz
Approx 45% complete for 20191112.A-G4_R1.fastq.gz
Approx 65% complete for 20191112.A-GE3_R1.fastq.gz
Approx 55% complete for 20191112.A-GE1_R1.fastq.gz
Approx 70% complete for 20191112.A-GE2_R1.fastq.gz
Approx 60% complete for 20191112.A-G2_R1.fastq.gz
Approx 45% complete for 20191112.A-G3_R1.fastq.gz
Approx 70% complete for 20191112.A-GE3_R1.fastq.gz
Approx 75% complete for 20191112.A-GE2_R1.fastq.gz
Approx 40% complete for 20191112.A-G1_R1.fastq.gz
Approx 60% complete for 20191112.A-GE1_R1.fastq.gz
Approx 50% complete for 20191112.A-G4_R1.fastq.gz
Approx 25% complete for 20191112.A-GE4_R1.fastq.gz
Approx 65% complete for 20191112.A-G2_R1.fastq.gz
Approx 80% complete for 20191112.A-GE2_R1.fastq.gz
Approx 75% complete for 20191112.A-GE3_R1.fastq.gz
Approx 50% complete for 20191112.A-G3_R1.fastq.gz
Approx 65% complete for 20191112.A-GE1_R1.fastq.gz
Approx 70% complete for 20191112.A-G2_R1.fastq.gz
Approx 55% complete for 20191112.A-G4_R1.fastq.gz
Approx 85% complete for 20191112.A-GE2_R1.fastq.gz
Approx 45% complete for 20191112.A-G1_R1.fastq.gz
Approx 80% complete for 20191112.A-GE3_R1.fastq.gz
Approx 70% complete for 20191112.A-GE1_R1.fastq.gz
Approx 55% complete for 20191112.A-G3_R1.fastq.gz
Approx 75% complete for 20191112.A-G2_R1.fastq.gz
Approx 90% complete for 20191112.A-GE2_R1.fastq.gz
Approx 85% complete for 20191112.A-GE3_R1.fastq.gz
Approx 60% complete for 20191112.A-G4_R1.fastq.gz
Approx 30% complete for 20191112.A-GE4_R1.fastq.gz
Approx 75% complete for 20191112.A-GE1_R1.fastq.gz
Approx 50% complete for 20191112.A-G1_R1.fastq.gz
Approx 80% complete for 20191112.A-G2_R1.fastq.gz
Approx 95% complete for 20191112.A-GE2_R1.fastq.gz
Approx 90% complete for 20191112.A-GE3_R1.fastq.gz
Approx 60% complete for 20191112.A-G3_R1.fastq.gz
Approx 65% complete for 20191112.A-G4_R1.fastq.gz
Analysis complete for 20191112.A-GE2_R1.fastq.gz
Approx 80% complete for 20191112.A-GE1_R1.fastq.gz
Approx 85% complete for 20191112.A-G2_R1.fastq.gz
Approx 95% complete for 20191112.A-GE3_R1.fastq.gz
Approx 55% complete for 20191112.A-G1_R1.fastq.gz
Approx 65% complete for 20191112.A-G3_R1.fastq.gz
Analysis complete for 20191112.A-GE3_R1.fastq.gz
Approx 90% complete for 20191112.A-G2_R1.fastq.gz
Approx 70% complete for 20191112.A-G4_R1.fastq.gz
Approx 85% complete for 20191112.A-GE1_R1.fastq.gz
Approx 35% complete for 20191112.A-GE4_R1.fastq.gz
Approx 70% complete for 20191112.A-G3_R1.fastq.gz
Approx 95% complete for 20191112.A-G2_R1.fastq.gz
Approx 90% complete for 20191112.A-GE1_R1.fastq.gz
Approx 60% complete for 20191112.A-G1_R1.fastq.gz
Approx 75% complete for 20191112.A-G4_R1.fastq.gz
Analysis complete for 20191112.A-G2_R1.fastq.gz
Approx 95% complete for 20191112.A-GE1_R1.fastq.gz
Approx 75% complete for 20191112.A-G3_R1.fastq.gz
Approx 80% complete for 20191112.A-G4_R1.fastq.gz
Approx 65% complete for 20191112.A-G1_R1.fastq.gz
Approx 40% complete for 20191112.A-GE4_R1.fastq.gz
Analysis complete for 20191112.A-GE1_R1.fastq.gz
Approx 80% complete for 20191112.A-G3_R1.fastq.gz
Approx 85% complete for 20191112.A-G4_R1.fastq.gz
Approx 70% complete for 20191112.A-G1_R1.fastq.gz
Approx 85% complete for 20191112.A-G3_R1.fastq.gz
Approx 90% complete for 20191112.A-G4_R1.fastq.gz
Approx 45% complete for 20191112.A-GE4_R1.fastq.gz
Approx 75% complete for 20191112.A-G1_R1.fastq.gz
Approx 95% complete for 20191112.A-G4_R1.fastq.gz
Approx 90% complete for 20191112.A-G3_R1.fastq.gz
Approx 80% complete for 20191112.A-G1_R1.fastq.gz
Analysis complete for 20191112.A-G4_R1.fastq.gz
Approx 50% complete for 20191112.A-GE4_R1.fastq.gz
Approx 95% complete for 20191112.A-G3_R1.fastq.gz
Approx 85% complete for 20191112.A-G1_R1.fastq.gz
Analysis complete for 20191112.A-G3_R1.fastq.gz
Approx 55% complete for 20191112.A-GE4_R1.fastq.gz
Approx 90% complete for 20191112.A-G1_R1.fastq.gz
Approx 95% complete for 20191112.A-G1_R1.fastq.gz
Approx 60% complete for 20191112.A-GE4_R1.fastq.gz
Analysis complete for 20191112.A-G1_R1.fastq.gz
Approx 65% complete for 20191112.A-GE4_R1.fastq.gz
Approx 70% complete for 20191112.A-GE4_R1.fastq.gz
Approx 75% complete for 20191112.A-GE4_R1.fastq.gz
Approx 80% complete for 20191112.A-GE4_R1.fastq.gz
Approx 85% complete for 20191112.A-GE4_R1.fastq.gz
Approx 90% complete for 20191112.A-GE4_R1.fastq.gz
Approx 95% complete for 20191112.A-GE4_R1.fastq.gz
Analysis complete for 20191112.A-GE4_R1.fastq.gz

  /// MultiQC üîç | v1.14

|           multiqc | Search path : /home/rstudio/work/rna-seq-course-feb2023/results/1_qc/fastqc
|         searching | ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ 100% 152/152  
|            fastqc | Found 8 reports
|           multiqc | Compressing plot data
|           multiqc | Report      : results/1_qc/multiqc/multiqc_report.html
|           multiqc | Data        : results/1_qc/multiqc/multiqc_data
|           multiqc | MultiQC complete</code></pre>
</div>
</div>
<p>Assuming we are not carrying on with manual execution of fastp in the next section, run the following block to delete the raw data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-rf</span> ../../data/tmp/yeast_fastq_full</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exersise-2" class="level2">
<h2 class="anchored" data-anchor-id="exersise-2">Exersise #2</h2>
<p>Take a look at the outputs in the ‚Äúresults/1_qc/fastqc‚Äù and ‚Äúresults/1_qc/multiqc‚Äù directory to answer the following questions.</p>
<ol type="1">
<li>Did the pooling work well?</li>
<li>What could be an explanation for the decrease in quality towards the ends of the reads?</li>
<li>How does the per-base sequence content look for each sample?</li>
<li>Do you think it makes sense to sequence longer based on these data?</li>
<li>What would be a good method of preprocessing?</li>
</ol>
<div class="callout-note callout callout-style-simple callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>There were most likely some problems with the pooling. GE4 appears to be over-represented while other samples are comparatively underrepresented. This could be due to the sequencer type: some Illumina instruments can prefer shorter fragments during bridge amplification.</li>
<li>This is a multi-faceted issue. It is a general artifact of Illumina sequencers on one hand, related towards the speed of sequencing and the longevity of the reagents. We sequenced on the NextSeq, which has a maximum read length of 150. On another, adapter content towards the 3‚Äô end of the read can negatively affect the performance.</li>
<li>It looks as expected. GE4 however has an increase in the amount of Gs towards the end of the read. This is most likely due to very short fragment lengths (or adapter dimers) and the sequencer beginning to read past the adapter.</li>
<li>No, the adapter content is already quite high for most of the samples. Sequencing longer fragments would not be worth the investment, most likely.</li>
<li>Trimming off the adapter and removing the first few bases off the 5‚Äô end would remove most of the marking of bad quality from these samples, especially GE4.</li>
</ol>
</div>
</div>
</div>
</section>
</section>
<section id="fastqscreen" class="level1">
<h1>FastqScreen</h1>
<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/_build/html/index.html">FastqScreen</a> is a tool which aligns fastq files to a database of different genomes to determine from where the data originate. This can be used for a variety of purposes, such as for metagenomic characterisation, but we will use it as a QC tool to determine any potential sources of contamination.</p>
<p>Unfortunately, executing <code>fastq_screen</code> can take a significant amount of time even on a larger amount of resources than we have available in the course, so we will simply look at the results. The application is installed, however, and you can take a look at the arguments it would take if we were to execute it ourselves.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/opt/fastq_screen_v0.14.0/fastq_screen</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="scrolling"><code>    
FastQ Screen - Map sequences against multiple genomes

www.bioinformatics.babraham.ac.uk/projects/fastq_screen
Contact: steven.wingett@babraham.ac.uk

Synopsis

  fastq_screen [OPTIONS]... [FastQ FILE]...

Function

  FastQ Screen is intended to be used as part of a QC pipeline.
  It allows you to take a sequence dataset and search it
  against a set of bowtie databases.  It will then generate
  both a text and a graphical summary of the results to see if
  the sequence dataset contains the kind of sequences you expect.

Options

 --add_genome &lt;text&gt;  Edits the file 'fastq_screen.conf' (in the folder where
                      this script is saved) to add a new genome. Specify the 
                      additional genome as a comma separated list:
                      'Database name','Genome path and basename','Notes'

 --aligner &lt;func&gt;     Specify the aligner to use for the mapping. Valid 
                      arguments are 'bowtie', bowtie2' (default) or 'bwa'.  
                      Bowtie maps with parameters -k 2, Bowtie 2 with 
                      parameters -k 2 --very-fast-local and BWA with mem -a.  
                      Local aligners such as BWA or Bowtie2 will be better 
                      at detecting the origin of chimeric reads. 

 --bisulfite          Process bisulfite libraries. The path to the 
                      bisulfite aligner (Bismark) may be specified in the 
                      configuration file. Bismark runs in non-directional 
                      mode. Either conventional or bisulfite libraries may
                      be processed, but not both simultaneously. The 
                      --bisulfite option cannot be used in conjunction with 
                      --bwa.

 --bowtie &lt;text&gt;      Specify extra parameters to be passed to Bowtie. 
                      These parameters should be quoted to clearly 
                      delimit bowtie parameters from fastq_screen 
                      parameters. You should not try to use this option 
                      to override the normal search or reporting options 
                      for bowtie which are set automatically but it might 
                      be useful to allow reads to be trimmed before
                      alignment etc.

 --bowtie2 &lt;text&gt;     Specify extra parameters to be passed to Bowtie 2. 
                      These parameters should be quoted to clearly 
                      delimit Bowtie 2 parameters from FastQ Screen 
                      parameters. You should not try to use this option 
                      to override the normal search or reporting options 
                      for bowtie which are set automatically but it might 
                      be useful to allow reads to be trimmed before
                      alignment etc.  

 --bwa &lt;text&gt;         Specify extra parameters to be passed to BWA. 
                      These parameters should be quoted to clearly 
                      delimit BWA parameters from FastQ Screen 
                      parameters. You should not try to use this option 
                      to override the normal search or reporting options 
                      for BWA which are set automatically but it might 
                      be useful to allow reads to be trimmed before
                      alignment etc. 

 --conf &lt;path&gt;        Manually specify a location for the configuration.
 
 --filter &lt;text&gt;      Produce a FASTQ file containing reads mapping to 
                      specified genomes. Pass the argument a string of
                      characters (0, 1, 2, 3, -), in which each character 
                      corresponds to a reference genome (in the order the 
                      reference genome occurs in the configuration file).  
                      Below gives an explanation of each character.     
                      0: Read does not map
                      1: Read maps uniquely
                      2: Read multi-maps
                      3: Read maps (one or more times)
                      4: Passes filter 0 or filter 1
                      5: Passes filter 0 or filter 2
                      -: Do not apply filter to this genome
                
                      Consider mapping to three genomes (A, B and C), the 
                      string '003' produces a file in which reads do not 
                      map to genomes A or B, but map (once or more) to 
                      genome C.  The string '--1' would generate a file in 
                      which reads uniquely map to genome C. Whether reads 
                      map to genome A or B would be ignored.
                      
                      A read needs to pass all the genome filters to be
                      considered valid (unless --pass specified).
               
                      When --filter is used in conjuction with --tag, FASTQ
                      files shall be mapped, tagged and then filtered. If
                      the --tag option is not selected however, the input 
                      FASTQ file should have been previously tagged.
                
 --force              Do not terminate if output files already exist,
                      instead overwrite the files.
 
 --get_genomes        Download pre-indexed Bowtie2 genomes for a range of
                      commonly studied species and sequences. If used with
                      --bisulfite, Bismark bisulfite Bowtie2 indices will
                      be downloaded instead.                      
                      
 --help               Print program help and exit.

 --illumina1_3        Assume that the quality values are in encoded in
                      Illumina v1.3 format. Defaults to Sanger format
                      if this flag is not specified.

 --inverse            Inverts the --filter results i.e. reads that pass
                      the --filter parameter will not pass when 
                      --filter --inverse are specified together, and vice
                      versa.                      

 --nohits             Writes to a file the sequences that did not map to 
                      any of the specified genomes. This option is 
                      equivalent to specifying --tag --filter 0000 (number
                      of zeros corresponds to the number of genomes
                      screened).  By default the whole input file will be
                      mapped, unless overridden by --subset.                

 --outdir &lt;text&gt;      Specify a directory in which to save output files.
                      If no directory is specified then output files
                      are saved in the current working directory.
                      
 --pass &lt;int&gt;         Used in conjunction with --filter. By default all
                      genome filters must be passed for a read to pass
                      the --filter option.  However, a minimum number 
                      of genome filters may be specified that a read
                      needs pass to be considered to pass the --filter
                      option. (--pass 1 effecitively acts as an OR
                      boolean operator for the genome filters.)                   
                      
 --quiet              Suppress all progress reports on stderr and only
                      report errors.

 --subset &lt;int&gt;       Don't use the whole sequence file, but create a 
                      temporary dataset of this specified number of 
                      reads. The dataset created will be of 
                      approximately (within a factor of 2) of this size. 
                      If the real dataset is smaller than twice the 
                      specified size then the whole dataset will be used. 
                      Subsets will be taken evenly from throughout the 
                      whole original dataset. By Default FastQ Screen 
                      runs with this parameter set to 100000. To process
                      an entire dataset however, adjust --subset to 0.

--tag                 Label each FASTQ read header with a tag listing to 
                      which genomes the read did, or did not align. The 
                      first read in the output FASTQ file will list the 
                      full genome names along with a score denoting 
                      whether the read did not align (0), aligned 
                      uniquely to the specified genome (1), or aligned 
                      more than once (2). In subsequent reads the 
                      genome names are omitted and only the score is 
                      printed, in the same order as the first line.

                      This option results in the he whole file being 
                      processed unless overridden explicitly by the user 
                      with the --subset parameter 

--threads &lt;int&gt;       Specify across how many threads bowtie will be
                      allowed to run. Overrides the default value set
                      in the configuration file

--top &lt;int&gt;/&lt;int,int&gt; Don't use the whole sequence file, but create a 
                      temporary dataset of the specified number of 
                      reads taken from the top of the original file. It is
                      also possible to specify the number of lines to skip
                      before beginning the selection e.g. 
                      --top 100000,5000000 skips the first five million 
                      reads and selects the subsequent one hundred thousand 
                      reads. While this option is usually faster than 
                      comparable --subset operations, it does not prevent 
                      biases arising from non-uniform distribution of 
                      reads in the original FastQ file. This option should 
                      only be used when minimising processing time is of 
                      highest priority. 

--version             Print the program version and exit.

2019 Babraham Institute, Cambridge, UK </code></pre>
</div>
</div>
<p>Let‚Äôs download these results now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Make directories</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="va">OUTPUT_DIR</span><span class="op">=</span>results/1_qc/fastq_screen</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUTPUT_DIR</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$OUTPUT_DIR</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://fgcz-gstore.uzh.ch/public/RNASeqCourse/yeast_fastq_screen.html</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Back home</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 6947k  100 6947k    0     0  32.4M      0 --:--:-- --:--:-- --:--:-- 32.4M</code></pre>
</div>
</div>
<p>Open the downloaded html file which has been placed within <code>results/1_qc/fastq_screen</code>.</p>
<section id="exercise-3" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3">Exercise #3</h2>
<ol type="1">
<li>Which sample has the lowest rate of alignment to adapter sequence? Which has the highest?</li>
<li>Which organism next to yeast has the highest hit for the samples?</li>
</ol>
<div class="callout-note callout callout-style-simple callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>GE3 has the lowest rate of alignment to adapter sequence while GE4 has the highest rate. This is consistent with the fastqc result as well, where we saw a similar pattern for GE4.</li>
<li>The second highest hit for all samples is Bacillus cereus. This is a bacterium found commonly in the environment and the hits here are in all likelihood not simply by chance. The mapping rate is fairly low so we don‚Äôt expect this degree of contamination to have much impact on the samples.</li>
</ol>
</div>
</div>
</div>
</section>
</section>
<section id="preprocessing-with-fastp" class="level1">
<h1>Preprocessing with <code>fastp</code></h1>
<p>In this section, we preprocess the fastqs to prepare them for alignment in the next chapter using fastp<span class="citation" data-cites="chen2018">(<a href="#ref-chen2018" role="doc-biblioref">Chen et al. 2018</a>)</span>. This command-line tool trims and quality-filters reads in the input fastqs and deposits the output into a new set of fastq files. fastp has many potential options that can be set depending on the requirements of the experiment, but we opt for fairly simple options here. Look over the script below to see what it does, but <strong>do not execute it</strong>. Depending on the resources, fastp can take a long time to execute for this number of samples (~1 hr in this case), so we opt to rather download the outputs for use in the next chapter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="va">FASTP_OUTDIR</span><span class="op">=</span>results/1_qc/fastp</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$FASTP_OUTDIR</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Start processing with fastp</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> FASTQ <span class="kw">in</span> data/tmp/yeast_fastq_full/<span class="pp">*</span>_R1.fastq.gz</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span> <span class="va">SAMPLE_NAME</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${FASTQ</span><span class="op">%</span>_R1.fastq.gz<span class="va">})</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="va">$SAMPLE_NAME</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">/opt/fastp</span> <span class="at">--in1</span> <span class="va">$FASTQ</span> <span class="dt">\</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">--out1</span> <span class="va">$FASTP_OUTDIR</span>/<span class="va">$SAMPLE_NAME</span><span class="st">"_trimmed_R1.fastq.gz"</span> <span class="dt">\</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">--thread</span> 4 <span class="at">--trim_front1</span> 4 <span class="at">--trim_tail1</span> 0 <span class="dt">\</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">--adapter_fasta</span> data/supplementary-files/allIllumina-forTrimmomatic-20160202.fa <span class="dt">\</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">--max_len1</span> 0 <span class="at">--max_len2</span> 0 <span class="at">--trim_poly_x</span> <span class="at">--average_qual</span> 20 <span class="dt">\</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">--poly_x_min_len</span> 10 <span class="at">--length_required</span> 30 <span class="at">--compression</span> 2 <span class="dt">\</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span><span class="op">&gt;</span> <span class="va">$FASTP_OUTDIR</span>/<span class="va">$SAMPLE_NAME</span><span class="st">"_preprocessing.log"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span> <span class="at">-f</span> <span class="va">$FASTQ</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Back home</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exercise-4" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4">Exercise #4</h2>
<ol type="1">
<li>What issues if any that you noticed in the FastQC output would be solved by using fastp with the arguments provided?</li>
<li>Which sequences are provided for trimming?</li>
<li>How could the sequence of operations of fastp affect trimming?</li>
</ol>
<div class="callout-note callout callout-style-simple callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>The adapter content in some of the samples</p></li>
<li><p>The file with all sequences considered to be adapter provided here <code>data/supplementary-files/allIllumina-forTrimmomatic-20160202.fa</code> . This list is not exhaustive and may need to be expanded if your library prep in question uses an adapter not in the list.</p></li>
<li><p>According to the documentation on GitHub, fastp performs trimming in the following order:</p>
<pre class="text"><code>1, UMI preprocessing (--umi)
2, global trimming at front (--trim_front)
3, global trimming at tail (--trim_tail)
4, quality pruning at 5' (--cut_front)
5, quality pruning by sliding window (--cut_right)
6, quality pruning at 3' (--cut_tail)
7, trim polyG (--trim_poly_g, enabled by default for NovaSeq/NextSeq data)
8, trim adapter by overlap analysis (enabled by default for PE data)
9, trim adapter by adapter sequence (--adapter_sequence, --adapter_sequence_r2. For PE data, this step is skipped if last step succeeded)
10, trim polyX (--trim_poly_x)
11, trim to max length (---max_len)</code></pre>
<p>This sequence of steps might not work for every application. As an example: If we know we need to trim bases after the 5‚Äô adapter, as might be the case in some types of library prep, one round of fastp processing will not be enough. We would need to run fastp once to perform adapter trimming and then run fastp again to trim off the remaining sequences at the 5‚Äô.</p></li>
</ol>
</div>
</div>
</div>

</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Andrews:2010tn" class="csl-entry" role="doc-biblioentry">
Andrews, S. 2010. <span>‚ÄúFASTQC. A Quality Control Tool for High Throughput Sequence Data.‚Äù</span>
</div>
<div id="ref-chen2018" class="csl-entry" role="doc-biblioentry">
Chen, Shifu, Yanqing Zhou, Yaru Chen, and Jia Gu. 2018. <span>‚ÄúFastp: An Ultra-Fast All-in-One FASTQ Preprocessor.‚Äù</span> <em>Bioinformatics</em> 34 (17): i884‚Äì90. <a href="https://doi.org/10.1093/bioinformatics/bty560">https://doi.org/10.1093/bioinformatics/bty560</a>.
</div>
<div id="ref-ewels2016" class="csl-entry" role="doc-biblioentry">
Ewels, Philip, M√•ns Magnusson, Sverker Lundin, and Max K√§ller. 2016. <span>‚ÄúMultiQC: Summarize Analysis Results for Multiple Tools and Samples in a Single Report.‚Äù</span> <em>Bioinformatics</em> 32 (19): 3047‚Äì48. <a href="https://doi.org/10.1093/bioinformatics/btw354">https://doi.org/10.1093/bioinformatics/btw354</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>