<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.613">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Part 4: Differential Expression and Pathway Enrichment</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="2_mapping_files/libs/clipboard/clipboard.min.js"></script>
<script src="2_mapping_files/libs/quarto-html/quarto.js"></script>
<script src="2_mapping_files/libs/quarto-html/popper.min.js"></script>
<script src="2_mapping_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="2_mapping_files/libs/quarto-html/anchor.min.js"></script>
<link href="2_mapping_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="2_mapping_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="2_mapping_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="2_mapping_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="2_mapping_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<style>
    .quarto-title-block .quarto-title-banner {
      background: #00A7FF;
    }
    </style>


<link rel="stylesheet" href="style.css">
</head>

<body>

<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Part 4: Differential Expression and Pathway Enrichment</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#mapping-with-star" id="toc-mapping-with-star" class="nav-link active" data-scroll-target="#mapping-with-star">Mapping with STAR</a>
  <ul class="collapse">
  <li><a href="#downloading-test-data" id="toc-downloading-test-data" class="nav-link" data-scroll-target="#downloading-test-data">Downloading test data</a>
  <ul class="collapse">
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1">Exercise #1</a></li>
  </ul></li>
  <li><a href="#running-star" id="toc-running-star" class="nav-link" data-scroll-target="#running-star">Running STAR</a>
  <ul class="collapse">
  <li><a href="#exercise-2" id="toc-exercise-2" class="nav-link" data-scroll-target="#exercise-2">Exercise #2</a></li>
  </ul></li>
  <li><a href="#alignment-qc" id="toc-alignment-qc" class="nav-link" data-scroll-target="#alignment-qc">Alignment QC</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="mapping-with-star" class="level1">
<h1>Mapping with STAR</h1>
<p>For this section, we will be aligning our processed fastqs using STAR <span class="citation" data-cites="dobin2012">(<a href="#ref-dobin2012" role="doc-biblioref">Dobin et al. 2012</a>)</span>.</p>
<section id="downloading-test-data" class="level2">
<h2 class="anchored" data-anchor-id="downloading-test-data">Downloading test data</h2>
<p>First, let’s download our dataset for this section. If you haven’t already, please delete all fastp generated data from the previous section. Don’t worry, we will download it at a later point again from the FGCZ server. Then, let’s download our test data for this section: fastp-processed files which have been downsampled by 10X to speed up alignment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the data</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="va">FASTP_OUTDIR</span><span class="op">=</span>results/1_qc/fastp_downsampled</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$FASTP_OUTDIR</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$FASTP_OUTDIR</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Depositing data in "</span><span class="va">$FASTP_OUTDIR</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://fgcz-gstore.uzh.ch/public/RNASeqCourse/yeast_fastp_downsampled_for_STAR.tar</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract data</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> yeast_fastp_downsampled_for_STAR.tar</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-f</span> yeast_fastp_downsampled_for_STAR.tar</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co"># back home</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Depositing data in results/1_qc/fastp_downsampled
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  1  518M    1 6224k    0     0  27.0M      0  0:00:19 --:--:--  0:00:19 26.8M
 20  518M   20  104M    0     0  85.4M      0  0:00:06  0:00:01  0:00:05 85.3M
 42  518M   42  223M    0     0   100M      0  0:00:05  0:00:02  0:00:03  100M
 67  518M   67  352M    0     0   109M      0  0:00:04  0:00:03  0:00:01  109M
 89  518M   89  466M    0     0   110M      0  0:00:04  0:00:04 --:--:--  110M
100  518M  100  518M    0     0   111M      0  0:00:04  0:00:04 --:--:--  115M
G1_trimmed_R1.fastq.gz
G2_trimmed_R1.fastq.gz
G3_trimmed_R1.fastq.gz
G4_trimmed_R1.fastq.gz
GE1_trimmed_R1.fastq.gz
GE2_trimmed_R1.fastq.gz
GE3_trimmed_R1.fastq.gz
GE4_trimmed_R1.fastq.gz</code></pre>
</div>
</div>
<p>Now that the data has been downloaded, let’s see exactly what we are dealing with. To this end, run the following cell. It calls <a href="https://bioinf.shenwei.me/seqkit/usage/">seqkit</a> to output some stats about the fastq files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/opt/seqkit</span> stats ../../results/1_qc/fastp_downsampled/<span class="pp">*</span>fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>file                                                          format  type   num_seqs      sum_len  min_len  avg_len  max_len
../../results/1_qc/fastp_downsampled/G1_trimmed_R1.fastq.gz   FASTQ   DNA   1,066,202  141,070,396       30    132.3      147
../../results/1_qc/fastp_downsampled/G2_trimmed_R1.fastq.gz   FASTQ   DNA     628,441   82,286,905       30    130.9      147
../../results/1_qc/fastp_downsampled/G3_trimmed_R1.fastq.gz   FASTQ   DNA     920,230  117,578,321       30    127.8      147
../../results/1_qc/fastp_downsampled/G4_trimmed_R1.fastq.gz   FASTQ   DNA     843,347  110,512,397       30      131      147
../../results/1_qc/fastp_downsampled/GE1_trimmed_R1.fastq.gz  FASTQ   DNA     710,351   93,240,086       30    131.3      147
../../results/1_qc/fastp_downsampled/GE2_trimmed_R1.fastq.gz  FASTQ   DNA     558,466   73,279,798       30    131.2      147
../../results/1_qc/fastp_downsampled/GE3_trimmed_R1.fastq.gz  FASTQ   DNA     583,743   78,807,718       30      135      147
../../results/1_qc/fastp_downsampled/GE4_trimmed_R1.fastq.gz  FASTQ   DNA   1,443,689  174,917,147       30    121.2      147</code></pre>
</div>
</div>
<section id="exercise-1" class="level3">
<h3 class="anchored" data-anchor-id="exercise-1">Exercise #1</h3>
<ol type="1">
<li><p>How many reads does each sample have? How does this compare to the input fastq files?</p></li>
<li><p>Is the minimum length of each read in the fastq files expected?</p></li>
</ol>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>As expected, each sample has around a tenth of the input reads of our initial sequencing data. 500’000 to 1’000’000 may seem like little considering the general recommendation of 20M for bulk RNA-seq, but it is enough to actually get reasonable results for yeast.</li>
<li>Yes, we specified a minimum length of 30bp when we ran fastp, which is what we see in the seqkit output.</li>
</ol>
</div>
</div>
</div>
</section>
</section>
<section id="running-star" class="level2">
<h2 class="anchored" data-anchor-id="running-star">Running STAR</h2>
<p>We are now ready to run the STAR aligner on our inputs. This part can take a while (~15 minutes), so look over commands in the meantime.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="va">CURR_DIR</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../..</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="va">TRIMMED_FASTQS</span><span class="op">=</span>results/1_qc/fastp_downsampled</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="va">OUTPUT_DIR</span><span class="op">=</span>results/2_mapping/STAR_alignment</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUTPUT_DIR</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Start processing with fastp</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> FASTQ <span class="kw">in</span> <span class="va">$TRIMMED_FASTQS</span>/<span class="pp">*</span>_trimmed_R1.fastq.gz</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">do</span> <span class="va">SAMPLE_NAME</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${FASTQ</span><span class="op">%</span>_trimmed_R1.fastq.gz<span class="va">})</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="bu">echo</span> <span class="va">$FASTQ</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># First do alignment with STAR using a pre-built index</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="bu">time</span> /opt/STAR <span class="dt">\</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genomeDir</span> data/supplementary-files/Ensembl_R64_genes_STARIndex <span class="dt">\</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outFileNamePrefix</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span><span class="st">"_"</span> <span class="dt">\</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesIn</span> <span class="va">$FASTQ</span> <span class="at">--twopassMode</span> None <span class="at">--runThreadN</span> 4 <span class="dt">\</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">--sjdbOverhang</span> 150 <span class="at">--outFilterType</span> BySJout <span class="at">--outFilterMatchNmin</span> 30 <span class="dt">\</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outFilterMismatchNmax</span> 10 <span class="at">--outFilterMismatchNoverLmax</span> 0.05 <span class="dt">\</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outMultimapperOrder</span> Random <span class="at">--alignSJDBoverhangMin</span> 1 <span class="dt">\</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">--alignSJoverhangMin</span> 8 <span class="at">--alignIntronMax</span> 1000 <span class="at">--alignMatesGapMax</span> 1000 <span class="dt">\</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outFilterMultimapNmax</span> 50 <span class="at">--chimSegmentMin</span> 15 <span class="dt">\</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    <span class="at">--chimJunctionOverhangMin</span> 15 <span class="at">--chimScoreMin</span> 15 <span class="at">--chimScoreSeparation</span> 10 <span class="dt">\</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outSAMstrandField</span> intronMotif <span class="at">--alignEndsProtrude</span> 3 ConcordantPair <span class="dt">\</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outSAMattributes</span> All <span class="at">--outStd</span> BAM_Unsorted <span class="at">--outSAMtype</span> BAM Unsorted <span class="dt">\</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outSAMattrRGline</span> ID:<span class="va">$SAMPLE_NAME</span> SM:<span class="va">$SAMPLE_NAME</span> <span class="dt">\</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>    <span class="at">--readFilesCommand</span> zcat <span class="op">&gt;</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.out.bam</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Sort and index</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  <span class="ex">samtools</span> sort <span class="at">-l</span> 9 <span class="at">-m</span> 2625M <span class="at">-@</span> 4 <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.out.bam <span class="at">-o</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.bam</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span> <span class="at">-f</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.out.bam</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>  <span class="ex">samtools</span> index <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.bam</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># “Guess” how RNA-seq sequencing were configured</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>  <span class="ex">infer_experiment.py</span> <span class="at">-r</span> data/supplementary-files/Ensembl_R64_genes/genes.bed <span class="dt">\</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span>.bam <span class="at">-s</span> 200000 <span class="op">&gt;</span> <span class="va">$OUTPUT_DIR</span>/<span class="va">$SAMPLE_NAME</span><span class="st">"_inferred.txt"</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-rf</span> <span class="va">$TRIMMED_FASTQS</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Back home</span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CURR_DIR</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>results/1_qc/fastp_downsampled/G1_trimmed_R1.fastq.gz

real    0m56.869s
user    1m32.094s
sys 0m3.575s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/G2_trimmed_R1.fastq.gz

real    0m33.619s
user    0m54.377s
sys 0m1.938s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/G3_trimmed_R1.fastq.gz

real    0m49.040s
user    1m20.951s
sys 0m2.545s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/G4_trimmed_R1.fastq.gz

real    0m46.886s
user    1m12.872s
sys 0m2.564s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/GE1_trimmed_R1.fastq.gz

real    0m37.790s
user    1m1.553s
sys 0m2.336s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/GE2_trimmed_R1.fastq.gz

real    0m30.103s
user    0m47.655s
sys 0m1.676s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/GE3_trimmed_R1.fastq.gz

real    0m32.177s
user    0m51.250s
sys 0m2.065s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled
results/1_qc/fastp_downsampled/GE4_trimmed_R1.fastq.gz

real    1m15.463s
user    2m7.279s
sys 0m3.380s
Reading reference gene model data/supplementary-files/Ensembl_R64_genes/genes.bed ... Done
Loading SAM/BAM file ...  Total 200000 usable reads were sampled</code></pre>
</div>
</div>
<p>Let’s first take a look at the outputs first. The above function ran STAR, sorted the bam output, indexed it, and ran a function to determine the strandedness of the inputs. Let’s perform a quick exercise related to the plain text outputs generated by STAR.</p>
<section id="exercise-2" class="level3">
<h3 class="anchored" data-anchor-id="exercise-2">Exercise #2</h3>
<ol type="1">
<li>Take a look at the files ending in “_Log.final.out”. Which sample has the highest percentage of primary alignments? Which has the lowest percentage?</li>
<li>What is the range of splices for the samples? Do you think these are real?</li>
<li>(Challenge) Find the answers to the above two questions programmatically.</li>
<li>We ran a python file called <code>infer_experiment.py</code>, which aims to infer the strandedness of bam alignments using a bed file for reference. These outputs were placed into the files ending in “_inferred.txt”. Is the predicted strandedness consistent with our expectations?</li>
</ol>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Let’s first answer number 3, as it will help us to answer the other questions. This is obviously not the only way to do this, so feel free to use your own (potenitally more efficient) solution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ ggplot2 3.4.0     ✔ purrr   1.0.1
✔ tibble  3.1.8     ✔ dplyr   1.1.0
✔ tidyr   1.3.0     ✔ stringr 1.5.0
✔ readr   2.1.2     ✔ forcats 0.5.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>starLogFiles <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="st">"../../results/2_mapping/STAR_alignment/*_Log.final.out"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>logFiles <span class="ot">&lt;-</span> <span class="fu">lapply</span>(starLogFiles, <span class="cf">function</span>(x) {</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(x, <span class="at">sep=</span><span class="st">"|"</span>, <span class="at">strip.white=</span><span class="cn">TRUE</span>, <span class="at">nrows=</span><span class="dv">9</span>, <span class="at">skip=</span><span class="dv">8</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>()</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  tmpVals <span class="ot">&lt;-</span> <span class="fu">str_trim</span>(tmp[,<span class="dv">2</span>])</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  tmpNames <span class="ot">&lt;-</span> tmp[,<span class="dv">1</span>]</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(tmpVals) <span class="ot">&lt;-</span> tmpNames</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">as_tibble</span>(<span class="fu">as.list</span>(tmpVals)))</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">%&gt;%</span> </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">`</span><span class="at">Uniquely mapped reads %</span><span class="st">`</span><span class="ot">=</span><span class="fu">sub</span>(<span class="st">"%"</span>, <span class="st">""</span>, <span class="st">`</span><span class="at">Uniquely mapped reads %</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_all</span>(., as.numeric)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>logFiles<span class="sc">$</span>Names <span class="ot">&lt;-</span> <span class="fu">basename</span>(starLogFiles)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(logFiles) <span class="co"># Solution to exercise 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Uniquely mapped reads number Uniquely mapped reads % Average mapped length
 Min.   : 507796              Min.   :87.78           Min.   :120.0        
 1st Qu.: 561694              1st Qu.:89.51           1st Qu.:128.4        
 Median : 705080              Median :90.72           Median :129.3        
 Mean   : 761372              Mean   :90.31           Mean   :128.4        
 3rd Qu.: 871211              3rd Qu.:91.20           3rd Qu.:129.9        
 Max.   :1283159              Max.   :91.99           Max.   :133.5        
 Number of splices: Total Number of splices: Annotated (sjdb)
 Min.   :2181             Min.   :1702                       
 1st Qu.:3894             1st Qu.:3405                       
 Median :5104             Median :4188                       
 Mean   :5269             Mean   :4425                       
 3rd Qu.:7055             3rd Qu.:5901                       
 Max.   :8108             Max.   :6585                       
 Number of splices: GT/AG Number of splices: GC/AG Number of splices: AT/AC
 Min.   :2124             Min.   :16.00            Min.   : 6.00           
 1st Qu.:3850             1st Qu.:27.00            1st Qu.:10.00           
 Median :5064             Median :32.50            Median :11.00           
 Mean   :5214             Mean   :37.12            Mean   :12.62           
 3rd Qu.:6986             3rd Qu.:42.50            3rd Qu.:13.50           
 Max.   :8004             Max.   :66.00            Max.   :27.00           
 Number of splices: Non-canonical    Names          
 Min.   : 1.000                   Length:8          
 1st Qu.: 1.750                   Class :character  
 Median : 2.500                   Mode  :character  
 Mean   : 5.375                                     
 3rd Qu.: 9.500                                     
 Max.   :14.000                                     </code></pre>
</div>
</div>
<ol type="1">
<li>G2 is the sample with the highest mapping rate. GE3 is the sample with the lowest mapping rate.</li>
<li>1’702 to 6’585. Only a fraction of these splice sites are expected to be real considering the low frequency of splicing in yeast.</li>
<li>(see code)</li>
<li>Yes, the output is largely antisense.</li>
</ol>
</div>
</div>
</div>
</section>
</section>
<section id="alignment-qc" class="level2">
<h2 class="anchored" data-anchor-id="alignment-qc">Alignment QC</h2>
<p>We have already performed a few QC steps in the above section. Let’s now explore these outputs a bit further using the python-package <a href="https://rseqc.sourceforge.net/">RSeQC</a>. This package contains a number of useful standalone python scripts for running general QC of alignment data, with an emphasis on RNAseq data. In fact, we already used one of these scripts above.</p>

</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-dobin2012" class="csl-entry" role="doc-biblioentry">
Dobin, Alexander, Carrie A. Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R. Gingeras. 2012. <span>“STAR: Ultrafast Universal RNA-Seq Aligner.”</span> <em>Bioinformatics</em> 29 (1): 15–21. <a href="https://doi.org/10.1093/bioinformatics/bts635">https://doi.org/10.1093/bioinformatics/bts635</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>